src/test/java/it/gov/pagopa/reward/service/counters/UserInitiativeCountersUnlockMediatorServiceTest.java

```diff
@@ -0,0 +1,91 @@
+package it.gov.pagopa.reward.service.counters;
+
+import it.gov.pagopa.common.utils.TestUtils;
+import it.gov.pagopa.reward.connector.repository.UserInitiativeCountersRepository;
+import it.gov.pagopa.reward.dto.trx.RewardTransactionDTO;
+import it.gov.pagopa.reward.model.counters.UserInitiativeCounters;
+import it.gov.pagopa.reward.service.RewardErrorNotifierService;
+import it.gov.pagopa.reward.test.fakers.RewardTransactionDTOFaker;
+import it.gov.pagopa.reward.utils.RewardConstants;
+import org.jetbrains.annotations.NotNull;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.springframework.kafka.support.Acknowledgment;
+import org.springframework.kafka.support.KafkaHeaders;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.support.MessageBuilder;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.eq;
+
+@ExtendWith(MockitoExtension.class)
+class UserInitiativeCountersUnlockMediatorServiceTest {
+
+    @Mock
+    UserInitiativeCountersRepository userInitiativeCountersRepositoryMock;
+
+    @Mock
+    RewardErrorNotifierService rewardErrorNotifierServiceMock;
+
+    UserInitiativeCountersUnlockMediatorService  userInitiativeCountersUnlockMediatorService;
+
+    @BeforeEach
+    void setUp() {
+        userInitiativeCountersUnlockMediatorService = new UserInitiativeCountersUnlockMediatorServiceImpl(
+                "appName",
+                userInitiativeCountersRepositoryMock,
+                rewardErrorNotifierServiceMock,
+                0L,
+                TestUtils.objectMapper);
+    }
+
+    @Test
+    void unexpectedJson() {
+        userInitiativeCountersUnlockMediatorService.execute(Flux
+                .just(getMessage("INVALID JSON")));
+
+        Mockito.verify(rewardErrorNotifierServiceMock, Mockito.times(1)).notifyTransactionResponse(any(), any(), eq(true), any());
+    }
+
+    @Test
+    void acceptedStateAuthorized() {
+        RewardTransactionDTO rewardTransactionDTO = RewardTransactionDTOFaker.mockInstance(1);
+        rewardTransactionDTO.setStatus(RewardConstants.PAYMENT_STATE_AUTHORIZED);
+
+        Mockito.when(userInitiativeCountersRepositoryMock.unlockPendingTrx(rewardTransactionDTO.getId()))
+                .thenReturn(Mono.just(new UserInitiativeCounters()));
+
+        userInitiativeCountersUnlockMediatorService.execute(Flux.just(getMessage(TestUtils.jsonSerializer(rewardTransactionDTO))));
```

Comment:
isn't kinda odd that we have to pass a message with Kafka headers and the serialized dto to test the business logic? We have a tight coupling among the communication technology (kafka), the data format exchange (json) and the business logic. 
It means the test must be rewritten if one of those external factors change.
May we try to find a safer way to test the service?

Reply:
the tested class extends the BaseKafkaConsumer class that explicitly treats the message as coming from kafka.

Reply:
is there any relevant advantage in the adoption of a custom kafkaConsumer over the default spring cloud kafka consumer like this [one](https://github.com/pagopa/idpay-wallet/blob/4526f93412112ea2a0e0203accdc6373b578b885/src/main/java/it/gov/pagopa/wallet/event/consumer/TransactionConsumer.java#L13)?

Reply:
yes, reactive behavior will disorder the sequence of the input messages: we are handling the right sequence order and then handling explicitly the commit at time and/or size threshold met

Reply:
ok, got it. Since we are using an additional layer upon kafka, it must be completely isolated from the business logic and therefore the layer tests and the service tests might be separated.
Visually:
```
service business logic
--------------------
baseConsumerKafka
--------------------
spring cloud stream
```
we will have a serious issue in code testability and isolation if the layers are entangled together, don't you agree @antonioT90  @GiovanaSolorzano ?

Reply:
the additional layer is implemented by the superclass BaseKafkaConsumer, its extensions contains just business logics.

The test could be written in order to call the method implemented by the extended class, instead of the public method exposed by the superclass, this way we will just test the business logics!

Is it ok for you @and-mora @GiovanaSolorzano ?

Reply:
yes, it sounds fine for testing purposes thank you

Reply:
@antonioT90 About design pattern instead, I didn't understand the Service interface, maybe you can give me an insight
```
/**
 *  This component will take a {@link RewardTransactionDTO} and unlock the {@link it.gov.pagopa.reward.model.counters.UserInitiativeCounters}
 * */
public interface UserInitiativeCountersUnlockMediatorService {
    void execute(Flux<Message<String>> transactionDTO);
}
```
The `execute()` method is already implemented by `BaseKafkaConsumer`. Should the business logic be contained there? (and wishfully with the `RewardTransactionDTO transactionDto` parameter).

Reply:
the public method is for the layer implemented by BaseKafkaConsumer, inside this abstract class there the overloaded method having directly the DTO

Reply:
ok, so the method with the DTO is inside the abstract class: i suppose you refer to `protected abstract Mono<R> execute(T payload, Message<String> message, Map<String, Object> ctx)`. Even here there are those 2 additional fields `Message<String> message, Map<String, Object> ctx` that are not related to business logic but sounds like internal details (kafka related).
If you want to test through this `execute()` method how you populate those fields?

Reply:
These fields are there just to support some potential useCases where those fields could be usefull (BaseKafkaConsumer has born as utility class to use as copy/paste or as utility library through the project):
* The Message could be usefull if you want to read some specific header or if you want to write the entire message
* The ctx map is used to have a variable where to store info to share between each step: it comes with the MSG_ID key where as default there is the entire payload, used by the performanceLog to print it together with timing information, the particular implementation could decide to print different info (the result as an example...)

Reply:
"Message" is not Kafka related: it belongs to Spring Integration; while the Map is the jdk class

Reply:
ok @antonioT90, not kafka related but still a strange form to communicate with a business oriented layer.

Since the business logic is not contained in the interface `execute()` method, the only purpose of the interface (and why it's implemented) is to externally expose that only one method. There is no polymorphism purpose and i could use directly the `serviceImpl` with no changes in the codebase, is it correct?

Reply:
The public `execute()` method is still to handle the Kafka logic: it cannot be invoked without configuring Kafka related metadata.

The protected `execute` is for the business logic: this function is accepting still a Message in order to give the access to the original Message: the actual implementation is using just the payload, thus it could be invoked from outside without caring about the Message.
Because it's protected, it could be invoke just from the class inside the same package.

We could move to a separate class the content of this protected method, fisically separating this 2 layers, allowing other class to invoke it if necessary

Reply:
ok, let's try to separate the layers and see how it works